R/geom_recessions.R

```diff
@@ -435,41 +458,63 @@ GeomRecessionsText <- ggproto(
   }
 
+  # best known URL for machine readable NBER file
   if (is_null(url)) {
-    url <- "https://www.nber.org/sites/default/files/2021-01/NBER%20chronology_062020.xlsx"
+    url <- "http://data.nber.org/data/cycles/cycle%20dates%20pasted.csv"
   }
 
   # locally bind variable names
-  start_char <- end_char <- start_date <- end_date <- NULL
-
-  temp.file <- paste(tempfile(),".xlsx",sep = "")
-  utils::download.file(url, temp.file, mode = "wb", quiet = quietly)
-
-  recessions <- readxl::read_excel(temp.file, skip = 2) %>%
-    # drop end matter
-    dplyr::slice(1:(n()-7)) %>%
-    # drop first row trough
-    dplyr::slice(-1) %>%
-    tibble::as_tibble() %>%
-    # rename character values
-    dplyr::rename(start_char = 1, end_char = 2) %>%
-    dplyr::mutate(
-      # convert character dates to R date
-      start_date = as.Date(stringr::str_replace(start_char, " ", " 1, "), format = "%B %d, %Y"),
-      end_date = as.Date(stringr::str_replace(end_char, " ", " 1, "), format = "%B %d, %Y"),
-      # convert R dates to numeric dates
-      start_num = lubridate::decimal_date(start_date),
-      end_num = lubridate::decimal_date(end_date)
-    ) %>%
-    dplyr::select(-3:-8)
-
-  message("Successfully fetched from NBER")
-
-  return(recessions)
+  start_char <- end_char <- start_date <- end_date <- ongoing <- index <- NULL
+
+  return(
+    tryCatch({
+      temp.file <- paste0(tempfile(),".csv")
+      utils::download.file(url, temp.file, mode = "wb", quiet = quietly)
+
+      recessions <- read.csv(temp.file) %>%
+        # drop first row trough
+        dplyr::slice(-1) %>%
+        tibble::as_tibble() %>%
+        # rename character values
+        dplyr::rename(start_char = 1, end_char = 2) %>%
+        dplyr::mutate(
+          # convert character dates to R date
+          start_date = as.Date(start_char),
+          end_date = as.Date(end_char)) %>%
+        dplyr::arrange(start_date) %>%
+        # Add a row number for identifying the last recession
+        mutate(index = row_number()) %>%
+        # Flag unfinished recessions
+        mutate(ongoing = case_when(
+          is.na(end_date) & index == max(.$index) ~ T,
+          TRUE ~ F
+        )) %>%
+        # If there is an ongoing recession in the last row, add January 2200 for
+        # graphing purposes
+        mutate(end_date = case_when(
+          ongoing ~ as.Date("2200-01-01"),
+          TRUE ~ end_date)) %>%
+        mutate(
+          # convert R dates to numeric dates
+          start_num = lubridate::decimal_date(start_date),
+          end_num = lubridate::decimal_date(end_date)) %>%
+        # eliminate index
+        select(-index)
+
+      if (!quietly) {message("Successfully fetched from NBER")}
+
+      # Return recessions
+      recessions
+    },
+    error = function(cond){
+      if (!quietly) message("WARMING: Fetch or processing failed. `NULL` returned.")
+      return(NA)

```

Comment:
If we're happy returning NA, the warning should say NA rather than NULL. However, I think NULL is probably the more accurate thing to return here. Also "warMing" :-)

Reply:
Good catch on both.